stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestions/SuggestionListController.kt

```diff
@@ -0,0 +1,71 @@
+package io.getstream.chat.android.ui.suggestions
+
+import android.widget.EditText
+import io.getstream.chat.android.client.models.Command
+import io.getstream.chat.android.client.models.Member
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.models.name
+import java.util.regex.Pattern
+
+internal class SuggestionListController(
+    private val suggestionListView: SuggestionListView,
+    private val messageTextInput: EditText
+) {
+    var members: List<Member> = emptyList()
+    var commands: List<Command> = emptyList()
+
+    private var messageText: String
+        get() = messageTextInput.text.toString()
+        set(text) {
+            messageTextInput.requestFocus()
+            messageTextInput.setText(text)
+            messageTextInput.setSelection(messageTextInput.text.length)
+        }
+
+    init {
+        suggestionListView.setOnSuggestionClickListener(
+            object : SuggestionListView.OnSuggestionClickListener {
+                override fun onMentionClick(user: User) {
+                    messageText = "${messageText.substringBeforeLast("@")}@${user.name} "
+                }
+
+                override fun onCommandClick(command: Command) {
+                    messageText = "/${command.name} "
+                }
+            }
+        )
+    }
+
+    fun onTextChanged(inputText: String) {
+        when {
+            inputText.isCommandMessage() -> suggestionListView.setSuggestions(inputText.getCommandSuggestions())
+            inputText.isMentionMessage() -> suggestionListView.setSuggestions(inputText.getMentionSuggestions())
+            else -> suggestionListView.clearSuggestions()
+        }
+    }
+
+    private fun String.isCommandMessage() = COMMAND_PATTERN.matcher(this).find()
+
+    private fun String.isMentionMessage() = MENTION_PATTERN.matcher(this).find()
+
+    private fun String.getCommandSuggestions(): SuggestionListView.Suggestions.CommandSuggestions {
+        val commandPattern = removePrefix("/")
+        return commands
+            .filter { it.name.startsWith(commandPattern) }
+            .toList()
+            .let { SuggestionListView.Suggestions.CommandSuggestions(it) }
+    }
+
+    private fun String.getMentionSuggestions(): SuggestionListView.Suggestions.MentionSuggestions {
+        val namePattern = substringAfterLast("@")
+        return members.map { it.user }
```

Comment:
We do this mapping every time that the input string changes, perhaps we should just store the mapped result to begin with (when `members` are set), so that we only compute it once

Reply:
Done